cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -711,6 +717,224 @@ class self_comparator {
+      it + 1, it + _table.num_columns(), initial_hash, [](auto hash, auto h) {
+        return cudf::detail::hash_combine(hash, h);
+      });
+  }
+
+ private:
+  /**
+   * @brief Computes the hash value of an element in the given column.
+   *
+   * When the column is non-nested, this is a simple wrapper around the element_hasher.
+   * When the column is nested, this uses the element_hasher to hash the shape and values of the
+   * column.
+   */
+  template <template <typename> class hash_fn>
+  class element_hasher_adapter {
+   public:
+    __device__ element_hasher_adapter(
+      Nullate nulls,
+      uint32_t seed             = DEFAULT_HASH_SEED,
+      hash_value_type null_hash = std::numeric_limits<hash_value_type>::max()) noexcept
+      : _has_nulls(nulls), _seed(seed), _null_hash(null_hash)
+    {
+    }
+
+    template <typename T, CUDF_ENABLE_IF(not cudf::is_nested<T>())>
+    __device__ hash_value_type operator()(column_device_view const& col,
+                                          size_type row_index) const noexcept
+    {
+      return element_hasher<hash_fn, Nullate>(_has_nulls, _seed, _null_hash)
+        .template operator()<T>(col, row_index);
+    }
+
+    template <typename T, CUDF_ENABLE_IF(cudf::is_nested<T>())>
+    __device__ hash_value_type operator()(column_device_view const& col,
+                                          size_type row_index) const noexcept
+    {
+      auto hash                   = hash_value_type{_seed};
+      column_device_view curr_col = col.slice(row_index, 1);
+      while (is_nested(curr_col.type())) {
+        if (_has_nulls) {
+          auto validity_it = detail::make_validity_iterator<true>(curr_col);
+          hash =
+            detail::accumulate(validity_it,
+                               validity_it + curr_col.size(),
+                               hash,
+                               [null_hash = _null_hash](auto hash, auto is_valid) {
+                                 return cudf::detail::hash_combine(hash, is_valid ? 0 : null_hash);
+                               });
+        }
+        if (curr_col.type().id() == type_id::STRUCT) {
+          if (curr_col.num_child_columns() == 0) { return hash; }
+          curr_col = detail::structs_column_device_view(curr_col).sliced_child(0);
+        } else if (curr_col.type().id() == type_id::LIST) {
+          auto list_col   = detail::lists_column_device_view(curr_col);
+          auto list_sizes = make_list_size_iterator(list_col);
+          hash            = detail::accumulate(
+            list_sizes, list_sizes + list_col.size(), hash, [](auto hash, auto size) {
+              return cudf::detail::hash_combine(hash, hash_fn<decltype(size)>{}(size));
+            });
+          curr_col = list_col.sliced_child();
+        }
+      }
+      for (int i = 0; i < curr_col.size(); ++i) {
+        hash = cudf::detail::hash_combine(
+          hash,
+          type_dispatcher<dispatch_void_if_nested>(
```

Comment:
Did you do any more analysis on why pulling the type dispatch out of the loop didn't help?

Reply:
Not analysis in the form of looking at cuobjdump or nsight but I did try throwing multiple tricks at it. Like within the range hasher, using a for loop was faster than this `detail::accumulate`. But changing any other usage of accumulate with a for loop didn't affect it.

Reply:
If someone more experienced could show me how to do said analysis it'd be very helpful. I haven't been able to get meaningful insights from our tools and I blame my own lack of knowledge of indicators.

Reply:
Frankly every time I use nsight compute it takes me _minimum_ 20-30 mins to get to a place where I have any idea what I'm doing, usually longer... Then it's 6 months+ before I need it again and I forget everything I learned the last time (so obviously I am not an expert).

What did you try? The usual results of type dispatcher issues are either high local memory traffic or in this case possibly low occupancy due to the amount of local state in this kernel with the nested types and such. `cuobjdump` could also let you look at register usage.

Reply:
Opened #10743 for future discussion.